File: 2016-10-26-CustomListViewSample/src/main/java/tech/thdev/customlistviewsample/network/RetrofitGitHub.java
package tech.thdev.customlistviewsample.network;

import retrofit2.Call;
import retrofit2.Retrofit;
import tech.thdev.customlistviewsample.data.GitHubUserResponse;

/**
 * Created by tae-hwan on 10/25/16.
 */

public class RetrofitGitHub {

    private GitHubService gitHubService;

    private static RetrofitGitHub retrofitGitHub;

    public static RetrofitGitHub getInstance() {
        if (retrofitGitHub == null) {
            synchronized (RetrofitGitHub.class) {
                if (retrofitGitHub == null) {
                    retrofitGitHub = new RetrofitGitHub();
                }
            }
        }
        return retrofitGitHub;
    }

    private RetrofitGitHub() {
        Retrofit retrofit = RetrofitCreator.createRetrofit();
        gitHubService = retrofit.create(GitHubService.class);
    }

    /**
     * Retrofit Search url
     */
    public Call<GitHubUserResponse> searchGitHubUser(String userKeyword, int page, int perPage) {
        return gitHubService.searchUser(userKeyword, page, perPage);
    }
}

File: app/src/main/java/tech/thdev/app/network/RetrofitGitHub.kt
package tech.thdev.app.network

import retrofit2.Call
import retrofit2.Retrofit
import tech.thdev.app.data.GitHubUserResponse

/**
 * Created by tae-hwan on 10/25/16.
 */
class RetrofitGitHub private constructor() {

    private val gitHubService: GitHubService

    init {
        val retrofit: Retrofit = createRetrofit()
        gitHubService = retrofit.create(GitHubService::class.java)
    }

    companion object {
        private var retrofitGitHub: RetrofitGitHub? = null
        fun newInstance(): RetrofitGitHub =
            retrofitGitHub ?: synchronized(RetrofitGitHub::class.java) {
                retrofitGitHub ?: RetrofitGitHub().also { retrofitGitHub = it }
            }
    }

    /**
     * Retrofit Search url
     */
    fun searchGitHubUser(
        userKeyword: String?,
        page: Int,
        perPage: Int
    ): Call<GitHubUserResponse> {
        return gitHubService.searchUser(userKeyword, page, perPage)
    }
}